core/devmode/src/main/java/io/quarkus/dev/RuntimeUpdatesProcessor.java

```diff
@@ -356,10 +372,34 @@ public void accept(Path path) {
         return ret;
     }
 
-    private boolean wasRecentlyModified(final Path p) {
+    private boolean sourceFileWasRecentModified(final Path sourcePath) {
+        return checkIfFileModified(sourcePath, sourceFileTimestamps);
+    }
+
+    private boolean classFileWasRecentModified(final Path classFilePath) {
+        return checkIfFileModified(classFilePath, classFileChangeTimestamps);
+    }
+
+    private boolean checkIfFileModified(Path path, Map<Path, Long> pathModificationTimes) {
         try {
-            long sourceMod = Files.getLastModifiedTime(p).toMillis();
-            return sourceMod > lastChange;
+            final long lastModificationTime = Files.getLastModifiedTime(path).toMillis();
+            final Long lastRecordedChange = pathModificationTimes.get(path);
+
+            if (lastRecordedChange == null) {
+                pathModificationTimes.put(path, lastModificationTime);
+                if (firstScanDone) {
+                    return true;
+                }
+
+                return false; // ignore changes to avoid recompilation and re-start
```

Comment:
I think this break the workflow of: Start Quarkus -> Make Change -> Make first request. This change will not be picked up because of the 'first scan' logic.

I don't really have a good solution for this, but I think that the above scenario will be much more common that the timestamps being wrong (once the issue is fixed in the code.quarkus.io generator).

Reply:
> I think this break the workflow of: Start Quarkus -> Make Change -> Make first request. This change will not be picked up because of the 'first scan' logic.
> 

Thanks Stuart for the review. Sure, we'll need two requests to make the changes in this workflow being taken into account. 

> I don't really have a good solution for this, but I think that the above scenario will be much more common that the timestamps being wrong (once the issue is fixed in the code.quarkus.io generator).

I was thinking of recording initial files timestamp in constructor. Will this work? WDYT?

Reply:
But we would need to wait for the thread to end before serving the first request, I suppose?

Reply:
I took a second look and this by trying to create a test case https://github.com/quarkusio/quarkus/pull/4135/files#diff-c0bdb27a09deae0802040076476bad74R250 for the scenario described above and this case seems already handled by the current code. 

I suppose with this line https://github.com/quarkusio/quarkus/blob/master/core/devmode/src/main/java/io/quarkus/dev/DevModeMain.java#L90 we'll record all the timestamps once a Quarkus app is started. All the modifications picked since the start and before the first call will be detected thus causing a re-compile since at this point we have done a first full scan and we have the timestamps recorded.

Reply:
Nice, in that case I think this is good to go.